packages/frontend/src/utils/major2-validation.ts

```diff
@@ -0,0 +1,672 @@
+  Course: "COURSE",
+  Range: "RANGE",
+  And: {
+    Type: "AND",
+    UnsatChild: "AND_UNSAT_CHILD",
+    NoSolution: "AND_NO_SOLUTION",
+  },
+  Or: "OR",
+  XofMany: "XOM",
+  Section: "SECTION",
+} as const;
+type ChildError = MajorValidationError & { childIndex: number };
+type CourseError = {
+  type: typeof MajorValidationErrorType.Course;
+  requiredCourse: string;
+};
+const CourseError = (c: IRequiredCourse): CourseError => ({
+  type: MajorValidationErrorType.Course,
+  requiredCourse: courseToString(c),
+});
+type AndError = {
+  type: typeof MajorValidationErrorType.And.Type;
+  error:
+    | {
+        type: typeof MajorValidationErrorType.And.UnsatChild;
+        childErrors: Array<ChildError>;
+      }
+    | {
+        type: typeof MajorValidationErrorType.And.NoSolution;
+        discoveredAtChild: number;
+      };
+};
+const AndErrorUnsatChild = (childErrors: Array<ChildError>): AndError => ({
+  type: MajorValidationErrorType.And.Type,
+  error: { type: MajorValidationErrorType.And.UnsatChild, childErrors },
+});
+export const AndErrorNoSolution = (idx: number): AndError => ({
+  type: MajorValidationErrorType.And.Type,
+  error: {
+    type: MajorValidationErrorType.And.NoSolution,
+    discoveredAtChild: idx,
+  },
+});
+type OrError = {
+  type: typeof MajorValidationErrorType.Or;
+  childErrors: Array<ChildError>;
+};
+const OrError = (childErrors: Array<ChildError>): OrError => ({
+  type: MajorValidationErrorType.Or,
+  childErrors,
+});
+type XOMError = {
+  type: typeof MajorValidationErrorType.XofMany;
+  childErrors: Array<ChildError>;
+  minRequiredCredits: number;
+  maxPossibleCredits: number;
+};
+const XOMError = (
+  r: IXofManyCourse,
+  childErrors: Array<ChildError>,
+  maxPossibleCredits: number
+): XOMError => ({
+  type: MajorValidationErrorType.XofMany,
+  childErrors,
+  minRequiredCredits: r.numCreditsMin,
+  maxPossibleCredits,
+});
+type SectionError = {
+  type: typeof MajorValidationErrorType.Section;
+  sectionTitle: string;
+  childErrors: Array<ChildError>;
+  minRequiredChildCount: number;
+  maxPossibleChildCount: number;
+};
+const SectionError = (
+  r: Section,
+  childErrors: Array<ChildError>,
+  max: number
+): SectionError => ({
+  type: MajorValidationErrorType.Section,
+  sectionTitle: r.title,
+  childErrors,
+  minRequiredChildCount: r.minRequirementCount,
+  maxPossibleChildCount: max,
+});
+type TotalCreditsRequirementError = {
+  takenCredits: number;
+  requiredCredits: number;
+};
+// handy result type, is either OK(T) or ERR(E)
+type Result<T, E> = { ok: T; type: "OK" } | { err: E; type: "ERR" };
+export const Ok = <T, E>(ok: T): Result<T, E> => ({ ok, type: "OK" });
+export const Err = <T, E>(err: E): Result<T, E> => ({ err, type: "ERR" });
+
+export const assertUnreachable = (_: never): never => {
+  throw new Error("This code is unreachable");
+};
+
+export const sectionToReq2 = (s: Section): Requirement2 => ({
+  ...s,
+  type: "SECTION",
+});
```

Comment:
ya thinking that if we have `type` in Section too then we don't need this function anymore

Reply:
yeet